modules/images/webp-uploads/load.php

```diff
@@ -157,6 +166,56 @@ function webp_uploads_create_sources_property( array $metadata, $attachment_id )
 
 add_filter( 'wp_generate_attachment_metadata', 'webp_uploads_create_sources_property', 10, 2 );
 
+/**
+ * Filter the image editor default output format mapping to select the most appropriate
+ * output format depending on desired output formats and supported mime types by the image
+ * editor.
+ *
+ * @since n.e.x.t
+ *
+ * @param string $output_format The image editor default output format mapping.
+ * @param string $filename      Path to the image.
+ * @param string $mime_type     The source image mime type.
+ * @return string The new output format mapping.
+ */
+function webp_uploads_filter_image_editor_output_format( $output_format, $filename, $mime_type ) {
+	// Skip conversion when creating the `-scaled` image (for large image uploads).
+	if ( preg_match( '/-scaled\..{3}.?$/', $filename ) ) {
+		return $output_format;
+	}
+
+	// Use the original mime type if this type is allowed.
+	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
+	if (
+		! isset( $valid_mime_transforms[ $mime_type ] ) ||
+		! is_array( $valid_mime_transforms[ $mime_type ] ) ||
+		in_array( $mime_type, $valid_mime_transforms[ $mime_type ], true )
+	) {
+		return $output_format;
+	}
```

Comment:
I think the condition is not required as we actually need to return the first element from the array instead.

```suggestion
	) {
		return $valid_mime_transforms[ $mime_type ][0];
	}
```

That would effectively use the first element of the array as the primary mime type so for example if you have:

```php
'image/jpeg' => array( 'image/webp', 'image/jpeg' );
```

It should create by default the images as WebP and the sources as JPEG instead. However, I'm thinking that returning the first element prevents other mime types from the list to be returned so in that scenario the filter should be removed from where the sources is created. 

Let me know what you think or maybe I may be missing something.

Reply:
Your suggestion will not work if the returned value from the `webp_uploads_get_supported_image_mime_transforms` function is not an array or is an empty array.

Reply:
- If `webp_uploads_get_supported_image_mime_transforms` return `null` this would not create the `sources` property which is expected in this case.
- If the provided mime type is invalid the code already handles this scenario here: https://github.com/WordPress/wordpress-develop/blob/f0dfa682a5cf2004f468c18ba8d5c87c463cdeea/src/wp-includes/class-wp-image-editor.php#L370 https://github.com/WordPress/performance/blob/87cc7f2dcc88d6f446a3a86a49ba22d65e85ce17/modules/images/webp-uploads/load.php#L335

Reply:
> If webp_uploads_get_supported_image_mime_transforms return array() this would not create the sources property which is expected in this case.

This function is not about `sources`, this function tells the image editor which format to use for the default subsizes. If the `webp_uploads_get_supported_image_mime_transforms` function returns an empty array then your suggested changes to return the first item from the mime transforms array will cause an error/warning because there is no elements with `0` key in that array.

The same will happen if someone unassigns the first mime type in a hook for our filter:

```php
add_filter( 'webp_uploads_upload_image_mime_transforms', function( $transorms ) {
    unset( $transforms[ 'image/webp' ][ array_search( 'image/webp', $transforms[ 'image/webp' ] ) );
    return $transforms;
} );

function webp_uploads_get_supported_image_mime_transforms() {
	$image_mime_transforms = array(
		'image/jpeg' => array( 'image/jpeg', 'image/webp' ),
		'image/webp' => array( 'image/webp', 'image/jpeg' ),
	);

	$transforms = (array) apply_filters( 'webp_uploads_upload_image_mime_transforms', $image_mime_transforms );

    ...
}

function webp_uploads_filter_image_editor_output_format( $output_format, $filename, $mime_type ) {
	// Use the original mime type if this type is allowed.
	$valid_mime_transforms = webp_uploads_get_supported_image_mime_transforms();
	if (
		! isset( $valid_mime_transforms[ $mime_type ] ) ||
		in_array( $mime_type, $valid_mime_transforms[ $mime_type ], true )
	) {
		// This will cause php warning for all images uploaded in webp format because it is unset in the hook:
		// ---> PHP Warning:  Undefined array key 0 ...
		return $valid_mime_transforms[ $mime_type ][0]; 
	}

	...
}
```

> If the provided mime type is invalid the code already handles this scenario here:

Once again the condition that you refer to will not be executed if we return the incorrect mime type here. Thus the image editor will have to fallback to the default mime type.

Reply:
Makes sense I was confusing it to a higher level before this point, thanks for the feedback.